### PR TITLE
Module namespace cleanup for v2.0

### PR DESCRIPTION
**Pull request by mariosasko**
This is an attempt to make the user-facing `datasets`' submodule namespace cleaner:

In particular, this PR does the following:
* removes the unused `zip_nested` and `flatten_nest_dict` and their accompanying tests
* removes `pyarrow` from the top-level namespace
* properly uses `__all__` and the `from  <module> import *` syntax to avoid importing the `<module>`'s submodules
* cleans up the `utils` namespace
* moves the `temp_seed` context manage from `datasets/utils/file_utils.py` to `datasets/utils/py_utils.py`

### CODE REVIEW

**Comment by severo** [comment]
will it solve https://github.com/huggingface/datasets-preview-backend/blob/4c542a74244045929615640ccbba5a902c344c5a/pyproject.toml#L85-L89?

**Comment by mariosasko** [comment]
@severo No, this PR doesn't fix that issue in the current state. We can fix it by adding `__all__` to `datasets/__init__.py` and `datasets/formatting/__init__.py`. However, this would require updating `__all__` for each new function/class definition, which could become cumbersome, and we can't do this dynamically because `mypy` is a static type checker.

@lhoestq @albertvillanova WDYT?

File: src/datasets/__init__.py [hunk 1]
```diff
@@ -20,36 +20,26 @@
...
     )
 
+SCRIPTS_VERSION = "master" if version.parse(__version__).is_devrelease else __version__
+
+del pyarrow
+del version
+
 from .arrow_dataset import Dataset, concatenate_datasets
-from .arrow_reader import ArrowReader, ReadInstruction
-from .arrow_writer import ArrowWriter
```
Comment (lhoestq) [inline src/datasets/__init__.py:35]: Here ArrowWriter and ArrowReader are removed from the top level module. I've seen that the lxmert example in `transformers` needs `datasets.ArrowWriter`, so I'm not sure if we should remove them

**Review by lhoestq — COMMENTED** [review]
Thank you ! This is much cleaner

File: src/datasets/search.py [hunk 2]
```diff
@@ -293,7 +295,9 @@ def add_vectors(
 
         # Add vectors
         logger.info(f"Adding {len(vectors)} vectors to the faiss index")
-        for i in utils.tqdm(range(0, len(vectors), batch_size), disable=not utils.is_progress_bar_enabled()):
+        for i in utils.tqdm_utils.tqdm(
```
[the remark this answers is not in the document]
Comment (mariosasko) [reply src/datasets/search.py:298]: Is it okay if I address this in a separate PR before the release (I'll make `disable_progress_bar` public again + add `enable_progress_bar`, `enable_caching` and `disable_caching`) to close https://github.com/huggingface/datasets/issues/3586?

[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/search.py:298]: Sure !

File: src/datasets/__init__.py (hunk 1 above)
Comment (lhoestq) [reply src/datasets/__init__.py:35]: Ok :)

**Review by lhoestq — APPROVED** [review]

**Comment by lhoestq** [comment]
Feel free to merge this one if it's good for you :)

**Pull request merged**